### DISCUSSION TITLE
espnow latency to long...

### DISCUSSION THREAD

**Discussion by mkoenings** [pschatzmann/arduino-audio-tools#348] [Q&A] [answered]
Hello, 
this is a follow up to my previous question [all in one (audiokit, faust, Encoder, espnow) ... too much ?](https://github.com/pschatzmann/arduino-audio-tools/discussions/337)
I have a working setup now, which streams from audiokit ( LyraT lineIn )-> Faust -> SBC encoding -> ESPNow. the sound data can be received and decoded on another audio board.
This works great, but... the latency is unfortunately quite bad.
I can measure a lagging from analog-in to analog-out of about 100msecs.  In the musical world, this is like a eternity, acceptable is less then 20 msecs, less than 10msecs is good. Its way better than Bluetooth, but unfortunately not good enough for musical applications.
So I tried to find the culprit for the delay. I removed Faust, and also the Encoder and went down to 8000 samples/16 bits. But no luck it does not make a noticable difference.
Is this the espnow protocol , which has only 250 bytes of payload ? There are some hints on the Espressif forums about certain settings, which can be set with the python menu tool.  https://github.com/espressif/esp-idf/issues/3238
Any ideas are appreciated!

Thanks, Michael

**Comment by pschatzmann** [comment, marked as answer]
I would play around with the I2S buffer size parameters and the copy size: https://github.com/pschatzmann/arduino-audio-tools/wiki/Performance
I am not sure how much lag is coming from the network...

**Comment by mkoenings** [comment]
cool! I will try that. Thanks again!

**Comment by mkoenings** [comment]
I tried to isolate the source of the latency. So I created a sketch with just one audiokit, input->output. I could not get the latency from in to out below 64msecs, which is not a good value. I was expecting less than 10msecs, since there isn't any transformation going on.
I changed the values for 
DEFAULT_BUFFER_SIZE (256)
I2S_BUFFER_SIZE (256)
I2S_BUFFER_COUNT (2)
The latency is best with a sample_rate of 48000, a lower value increases the latency.
Any ideas on how I can improve that? Is it the ES8388 that cannot do better ?

I use a oszilloscope to measure the lag between the analog input and analog output, using a short klick sound.
I did the same measurement on a daisy board, it has a latency of about 4msecs.
Thanks!

**Comment by pschatzmann** [comment]
Did you use an Audiokit (with an ESP32) or a ESP8388 ?

Google provided me with the following:
- daisy board: STM32 ARM Cortex-M7 running at 480MHz internal 32-bit floating point processing 
- ESP32 runs at 240mhz support for 32-bit floating point unit 
- ESP8266 is 80 MHZ  no fpu

I would suggest that you try the excercise also with an ESP32, but I don't think that you will get below 10msecs on such a cheap platform and you will never get the floating point performance of a M7. 

The ESP32 should be similar with an STM32M4

I try to avoid floating point operations, but this is not always possible. Just this week I did some [performance comparison for the STK framework](https://github.com/pschatzmann/Arduino-STK/wiki/Performance) which is based on floats

**Comment by mkoenings** [comment]
ok, I get the (floating-) point :-)
I forgot to mention, I am using esp32 Wrover E (Lyra_T 4.3) running on 240mhz. There is no complex operation going on here, just the I2S for input and output. Adding the ( much more complex?) Faust library actually does not add much extra time, I tried that as well.
So I wonder if the ES8388 is the slow one here.
BTW: the guys here have a nice guitar pedal platform based on the ESP32-A1S https://www.deeptronic.com/blackstomp/
thanks, Michael

here is the code:



#include <AudioKitHAL.h>
#include <AudioKitSettings.h>
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "esp32-hal-cpu.h"



AudioKitStream kit_in; 


StreamCopy copier(kit_in, kit_in, 128);  // copy data from kit to kit

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
AudioLogger::instance().begin(Serial, AudioLogger::Error);

  auto cfg = kit_in.defaultConfig();
  cfg.sample_rate = 32000; 
  cfg.channels = 1;
  cfg.bits_per_sample = 16;
  cfg.buffer_count = 2;
  cfg.buffer_size = 128;
  
  // start I2S
  auto cfg_i2s = kit_in.defaultConfig(RXTX_MODE); 
  cfg_i2s.sample_rate = cfg.sample_rate; 
  cfg_i2s.channels = cfg.channels;
  cfg_i2s.bits_per_sample = cfg.bits_per_sample;
  cfg_i2s.input_device = AUDIO_HAL_ADC_INPUT_LINE1;
  //cfg_i2s.buffer_count = 2;
  //cfg_i2s.buffer_size = 128;
  
  
 
  
  kit_in.begin(cfg_i2s);
  kit_in.setVolume(80);
  
   Serial.println();
  Serial.print("CPU Frequency is: ");
  Serial.print(getCpuFrequencyMhz()); //Get CPU clock
  Serial.print(" Mhz");
  Serial.println();
  
}

// Arduino loop - copy sound to kit 
void loop() {

  copier.copy();
  
}

**Comment by pschatzmann** [comment]
Hmm, If you have a regular ESP32 you could test the latency added by the ES8388 by just sending the output to AnalogAudioStream.

Via google I found some other projects that indicate that the latency should be just a few ms. So there is definitely something strange going on! See https://github.com/garygru/yummyDSP - but that was for a AC101...

Oh...I think you messed up your code a bit. Delete the invalid and not used auto cfg = kit_in.defaultConfig(); and adjust the parameters on cfg_i2s

**Comment by mkoenings** [comment]
oh, yes thanks , well spotted. Too much editing..
I will dig deeper, Thanks a lot!
Michael

**Comment by mkoenings** [comment]
I think I found an explanation for the 'strange' behaviour of the latency.
in AudioKitHAL.h  line 70,61 :
  ```
int buffer_count = 6;
int buffer_size = 512;
```
these values are not overwritten by the macros for the  i2S buffers.
I can get a much better latency by reducing these values :-)

**Comment by pschatzmann** [comment]
I never noticed: This is definitely a bug!
I just committed a correction...

Will you share your results ?

**Comment by mkoenings** [comment]
I did not try many combinations of config values. With 32000 samples, 16bit , a buffersize of 128 and a buffer_count of 2 I could get 8msecs end to end. That's hardly noticable, very good. This is using am AudioKit for input and output. I am not sure if this takes 2 dma buffer lines then? 
Thanks!

**Comment by pschatzmann** [comment]
Oh, cool! I did not know that it is possible to go that low!
I think buffer_count==2 would just correspond to a double buffering (which is usually used on the STM32).

The current I2S settings are mostly driven by the internet radio streaming which is more reliable, the bigger the buffer.
